port async unlinked drain from illumos-nexenta

alek-p opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This patch is an async implementation of the existing sync `zfs_unlinked_drain()` function. This function is called at mount time and is responsible for freeing znodes that we didn't get to freeing before.
We don't have to hold mounting of the dataset until the unlinked list is fully drained as is done now. Since we can process the unlinked set asynchronously this results in a better user experience when mounting a dataset with entries in the unlinked set.

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The problem this patch is attempting to solve is long mount times when the dataset being mounted has a long list of znodes in the unlinked set.
Originally this issue came to us @datto as a request to investigate why a particular snapshot took a lot longer to create a clone from than the other snapshots. Looking at what the "stuck" clone was doing we saw variations of:
```
[<ffffffffc05a5b63>] cv_wait_common+0xb3/0x130 [spl]
[<ffffffffc05a5c38>] __cv_wait_io+0x18/0x20 [spl]
[<ffffffffc10f2463>] zio_wait+0x103/0x1e0 [zfs]
[<ffffffffc1066e5f>] dmu_tx_hold_free+0x1ff/0x900 [zfs]
[<ffffffffc1056d42>] dmu_free_long_range+0x192/0x270 [zfs]
[<ffffffffc10c878a>] zfs_rmnode+0x28a/0x350 [zfs]
[<ffffffffc10e944b>] zfs_zinactive+0xbb/0xd0 [zfs]
[<ffffffffc10e29c1>] zfs_inactive+0x61/0x250 [zfs]
[<ffffffffc10f9933>] zpl_evict_inode+0x43/0x60 [zfs]
[<ffffffff812314c1>] evict+0xc1/0x190
[<ffffffff812317a7>] iput+0x1c7/0x250
[<ffffffffc10c78f4>] zfs_unlinked_drain+0xb4/0xe0 [zfs]
[<ffffffffc10dabad>] zfs_sb_setup+0xed/0x160 [zfs]
[<ffffffffc10dc00f>] zfs_domount+0x29f/0x330 [zfs]
[<ffffffffc10f9a7c>] zpl_fill_super+0x2c/0x40 [zfs]
[<ffffffff81218a4f>] mount_nodev+0x4f/0xa0
[<ffffffffc10f9eef>] zpl_mount+0x4f/0x80 [zfs]
[<ffffffff8121974d>] mount_fs+0x3d/0x170
[<ffffffff81236187>] vfs_kern_mount+0x67/0x110
[<ffffffff8123882f>] do_mount+0x25f/0xd90
[<ffffffff8123969f>] SyS_mount+0x9f/0x100
[<ffffffff8185328e>] entry_SYSCALL_64_fastpath+0x22/0xc1
[<ffffffffffffffff>] 0xffffffffffffffff
```

<!-- If it fixes an open issue, please link to the issue here. -->

### Description
As can be seen above the reason that the `zfs_unlinked_drain()` call can take so long is that it eventually calls `iput()` which destroys/frees the corresponding inode and could sleep:
```
Puts an inode, dropping its usage count. If the inode use count hits zero, the inode is then freed and may also be destroyed.
Consequently, iput can sleep.
```
If there are other entires in the unlinked set this will loop untill all the znodes are freed synchronously continuing to hold up mount in the process.

To solve this problems and facilitate async unlinked list drain we added a pool-wide taskq for processing unlinked drain asyncronsoly. This patch tracks the state of the unlinked drain per `zfsvfs` and is able to stop async processing of the unlinked set if we need to unmount or suspend the filesystem.


This patch was inspired by work done @nexenta mostly by @skiselkov and @Ramzec.
The general approach is simular to a mix of code from the following patches https://github.com/Nexenta/illumos-nexenta/commit/ea0ae4efc65a839131 and https://github.com/Nexenta/illumos-nexenta/commit/5d09257a00d903.


<!-- Describe your changes in detail -->

### How Has This Been Tested?
I unit tested this by comparing clone times on a snapshot that was created immediately after several large files (~250GB) were deleted. The difference between the two mount times was ~8x. I expect the difference will be bigger for non-SSD systems.

The zfs-test looked good on my workstation as well.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented on include/sys/zfs_vfsops.h:
```diff
@@ -132,6 +140,11 @@ struct zfsvfs {
+	kmutex_t	z_drain_lock;
+	kcondvar_t	z_drain_cv;
+	drain_state_t	z_drain_state;
+	boolean_t	z_drain_continue;
```
This all feels a bit overkill.  What do you think about simply adding a `boolean_t z_draining;` with the other booleans above, and relying on the existing `z_lock` to protect it.  It looks like you also don't really need to add a cv and could simply use `task_wait()` to determine when all the pending work has completed.

alek-p replied:
Since we want to be able to cancel an individual (running) task for example - using `taskq_wait()` as implemented would not work since it's a pool-wide taskq.
It's possible to get away from this locking and use `taskq_wait()` if I use a dedicated per zfsvfs taskq but not sure that's a good idea?

behlendorf replied:
If you were to store the dispatched `taskqid` in the `zfsvfs` you could wait for that specific task using `taskq_wait_id()`.   Cancelling it is also possible with `taskq_cancel_id()` but setting the shutdown flag is nicer.  The only complication would be that these are both Linux specific extensions to the taskq interface.

alek-p replied:
got it, will do. Upstream should probably use Nexenta's async drain approach anyway since they have to deals with forced unmounts

alek-p commented on module/zfs/zfs_dir.c:
```diff
@@ -508,8 +510,55 @@ zfs_unlinked_drain(zfsvfs_t *zfsvfs)
+	if (taskq_dispatch(dsl_pool_unlinked_drain_taskq(
+	    dmu_objset_pool(zfsvfs->z_os)),
+	    (void (*)(void *))zfs_unlinked_drain_task, zfsvfs,
+	    TQ_NOSLEEP) == TASKQID_INVALID) {
```
The problem here is that draining is not interruptable until the task is running. My thought was that if we can't get the task to run right away I would rather have it fail and do a sync unlink drain than not schedule the task, and as a result, hold the unmount of the dataset.

I guess it doesn't really matter since if we're out of memory there are going to be other more serious problems.

behlendorf replied:
My concern is that using `TQ_NOSLEEP` makes a failure more likely since it disables the direct memory reclaim call paths, where `TQ_SLEEP` will wait for the needed memory.  Though as you say it probably doesn't matter to much since we're only talking about a handful of bytes here for the taskq entry, so any failure is unlikely.  But I do think `TQ_SLEEP` is slightly better since if the kernel can't give us a handful of bytes we're in a very bad spot.

alek-p commented:
Thanks for the reviews! I've started on the updated patch and hope to have it ready next week.

alek-p commented:
This version of the patch addresses most of the review comments. The parts I left out is the periodic flush of the unlinked set and the byte counters for freeing from the unlinked set.
I think adding an on-demand flush command or the infrastructure to flush on a timed interval could be valuable but since we now flush the unlinked set when we resume a suspended filesystem (so after recv and rollback) these commands could be used as a workarounds to drain the unlinked set if needed on demand.
The kstats I've added track how many entries are currently in the unlinked set and how many times we called the unlinked set add function. The later kstat may not be very useful since we always add files that are about to be free to the unlinked set. I could find a good way to show how many bytes are being processed in the unlinked set. I can revisit that if the added kstats don't provide enough visibility. 
Please let me know what you think of the updated patch.

behlendorf commented on include/sys/dataset_kstats.h:
```diff
@@ -43,6 +46,8 @@ typedef struct dataset_kstat_values {
 	kstat_named_t dkv_reads;
 	kstat_named_t dkv_nread;
+	kstat_named_t dkv_nunlinks;   /* num znodes added to unlinked set */
+	kstat_named_t dkv_nunlinked; /* num znodes processed by unlinked set */
```
After giving this some more thought, I think it would be a little more useful and consistent if these were both monotonically increasing counters.  On mount `dkv_nunlinks` would be initialized to the unlinked set size and incremented whenever a new entry was added to the unlinked set.  `dkv_nunlinked` would be initialized to zero on mount and incremented when an entry was removed from the unlinked set.  The difference between the counters would still allow you to calculate the unlinked set size.  Plus you'd have the raw data to stuff in a database if you liked, or calculate things like the async unlink rate.

behlendorf commented on tests/zfs-tests/tests/functional/mount/umount_unlinked_drain.ksh:
```diff
@@ -0,0 +1,119 @@
+
+
+UNLINK_SP_PARAM=/sys/module/zfs/parameters/zfs_unlink_suspend_progress
+default_unlink_sp=$(cat $UNLINK_SP_PARAM)
```
We want to use the helper function `default_unlink_sp=$(get_tunable zfs_unlink_suspend_progress)` for portability.

behlendorf commented on tests/zfs-tests/tests/functional/mount/umount_unlinked_drain.ksh:
```diff
@@ -0,0 +1,119 @@
+		for fn in $(seq 1 100); do
+			log_must eval "rm $TESTDIR.$fs/file-$fn &"
+		done
+		sleep 10
```
I'd think you can just use `wait` here to delay until all the background `rm`s complete.

behlendorf commented on tests/zfs-tests/tests/functional/mount/umount_unlinked_drain.ksh:
```diff
@@ -0,0 +1,119 @@
+		log_must mounted $TESTDIR.$fs
+		log_must unlinked_size_is 100 $TESTPOOL $TESTPOOL/$TESTFS.$fs
+
+		# confirm we ca drain and add to unlinked set at the same time
```
s/ca/can

ahrens commented:
I'd like to review this but I probably won't have time to until after the new year. Please wait for my review if possible.

ahrens commented:
@lundman didn't you implement something like this for OSX?  Any thoughts on this implementation?

ahrens commented on module/zfs/zfs_vfsops.c:
```diff
@@ -1888,6 +1904,7 @@ zfs_preumount(struct super_block *sb)
 
 	/* zfsvfs is NULL when zfs_domount fails during mount */
 	if (zfsvfs) {
+		zfs_unlinked_drain_stop_wait(zfsvfs);
```
I couldn't figure out when zfs_preumount() is called.  Is it possible that we later decide not to unmount it, in which case how do we restart the unlinked_drain task?

alek-p replied:
it's called in `zpl_kill_sb()`. That function is called by the Linux kernel on unmount so afaik it's not possible to later decide not to unmount.

behlendorf replied:
Sorry for the confusion, when `zfs_preumount()` is called all inodes and dentrys have **not** yet been freed as correctly described in the comment above `zfs_preumount()`.  The dentry and inode caches will be torn down immediately after `zfs_preumount()` exits when `zpl_kill_sb()` calls `kill_anon_super()`.

```c
static void
zpl_kill_sb(struct super_block *sb)
{
        zfs_preumount(sb);
        kill_anon_super(sb);
}
```

alek-p commented on module/zfs/zfs_dir.c:
```diff
@@ -508,8 +513,51 @@ zfs_unlinked_drain(zfsvfs_t *zfsvfs)
+	if (zfsvfs->z_draining) {
+		zfsvfs->z_drain_continue = B_FALSE;
+		taskq_wait_id(dsl_pool_unlinked_drain_taskq(
+		    dmu_objset_pool(zfsvfs->z_os)), zfsvfs->z_drain_task);
```
yeah, looks like a pool-wide taskq is not going to cut it here since if there are lots of tasks in the queue we could be stuck in the back and holding up unmount as a result. I'll see what other options make sense.

As for `iput()` my understanding is in-line with yours - it could take arbitrarily long for large files. There is no getting away from this though - we either wait before we can mount (as done currently) or we have to wait on unmount (like in this patch). The benefit of canceling is that we don't have to wait for ALL the entries in the unlinked set, but only the one that was processing during the unmount.

alek-p replied:
https://github.com/Nexenta/illumos-nexenta/commit/ea0ae4efc65a839131#diff-6d1e73746c8c45ebbc788dd57707c6c2R1888 and https://github.com/Nexenta/illumos-nexenta/commit/5d09257a00d903#diff-26f5af57e103cb3660f924e73c01dc2cR544 are a couple of places where ZoL differs there were more but these are the easiest to find

alek-p commented on module/zfs/zfs_vfsops.c:
```diff
@@ -1206,14 +1210,27 @@ zfsvfs_setup(zfsvfs_t *zfsvfs, boolean_t mounting)
+			if (zap_get_stats(zfsvfs->z_os, zfsvfs->z_unlinkedobj,
+			    &zs) == 0) {
+				dataset_kstats_update_nunlinks_kstat(
+				    &zfsvfs->z_kstat, zs.zs_num_entries);
```
Is there a common place to document kstats? The only kstat documentation in this patch is the comments in code and the discussion is this PR: https://github.com/zfsonlinux/zfs/pull/8142#discussion_r239973107

behlendorf replied:
There should be, but currently there isn't.  But I agree we should clearly explain them in a code comment so when these all get documented somewhere what they mean is clear.  Suggestions for better names welcome.

alek-p commented:
I've addressed most of the review comments in the latest patch. The biggest change is switching to `taskq_cancel_id()` which allows us to still use a pool wide unlink taskq while making sure we won't make unmount wait behind other potentially unrelated tasks. This is ready for another review.

ahrens commented on module/zfs/zfs_dir.c:
```diff
@@ -508,8 +513,51 @@ zfs_unlinked_drain(zfsvfs_t *zfsvfs)
 
 		zp->z_unlinked = B_TRUE;
 		iput(ZTOI(zp));
```
cool, I didn't realize that.

ahrens replied:
Thanks.  Canceling would be nice, but even without canceling this is an improvement over the current method, and it would be good enough nearly all the time (i.e. where you aren't trying to unmount it before the queue is drained).  AFAIK the typical use case is that you delete a large file, and then the system crashes and when you come back up, the mount "hangs" for hours while the large file reclamation completes.

lundman commented on module/zfs/dsl_pool.c:
```diff
@@ -223,6 +223,9 @@ dsl_pool_open_impl(spa_t *spa, uint64_t txg)
 	    max_ncpus * 8, INT_MAX, TASKQ_PREPOPULATE | TASKQ_DYNAMIC);
+	dp->dp_unlinked_drain_taskq = taskq_create("z_unlinked_drain",
+	    max_ncpus, defclsyspri, max_ncpus, INT_MAX,
+	    TASKQ_PREPOPULATE | TASKQ_DYNAMIC);
```
We did - as we'd run without unlinked-drain for the first release, and people had entries in the millions. But we simply spawned a thread to handle it, it was taken out again once users upgraded version.

lundman commented on tests/zfs-tests/tests/functional/mount/umount_unlinked_drain.ksh:
```diff
@@ -0,0 +1,119 @@
+
+function unlinked_size_is
+{
+	MAX_ITERS=5 # iteration to do before we consider reported number stable
```
Isn't `iput` just `VN_RELE` and what we usually use ? Could we have that instead? Or, since ZOL is taking over, should I add `iput` and `ZTOI` macros?

alek-p replied:
I originally had a `sleep 10` here but I belive this version is actually more robust than having a static sleep or TXGs threshold.

This code provides us with dynamic behavior since we keep looping until the unlinked set size has not changed for the last 5 loop iterations. That allows us to avoid unecessary test delays while still trying to ensure that the unlink set processing has "finished".

Having said that, it's easy for me to remove the loop and put back the `sleep 10` if you feel that is more appropriate.

behlendorf replied:
Adding a brief delay between iterations (say 0.1) might be useful, my only concern would be that if the checks are done to quickly we might be mistaken about it stabilizing.

ahrens replied:
From looking at the code, I couldn't figure out why we would ever do more than one iteration.  Why wouldn't the unlinked size be the expected size right away?

alek-p replied:
Since the test does the file deletions in a forked process my concern was that we may not have processed all of the backgrounded file deletions by the time we go to check the unlinked set size.
So perhaps I just put a `sleep 10` after the `rm` calls instead of doing this function loop

behlendorf replied:
I'd like to avoid adding these kinds of gratuitous sleeps, the ZTS runs long enough already and there's still no assurance any time you pick with be long enough.  The existing approach seems more robust.

ahrens replied:
Ah, I didn't see that you were running `rm` in the background.  How does that interact with `log_must`?  Can it actually log the output and check the return code after it's in the background?

Could we avoid the loop/sleep by waiting for the background `rm`'s to complete?  Normally `wait` would do that, but I don't know what the interaction is with `log_must eval`.

ahrens replied:
Alternatively, do we really need to do the 100 `rm`'s in the background?  I would guess that the fork/exec/startup overhead is much more than the actual `unlink` syscall.  You could probably improve performance by doing `log_must rm $TESTDIR.$fs/file-*` in the foreground.

alek-p replied:
I could have done `log_must rm $TESTDIR.$fs/file-*` sequentially in the foreground but I thought it was a better test to do it in "parallel".

alek-p commented:
After a discussion @datto we confirmed we do want to have draining canceling. I going to look at making this cross-platform by just spawning new threads to process the unlinked set.

alek-p commented:
In the latest version of this patch, I've added the comment for `iput()` that Matt requested.
I've also looked at making this cross-platform but could not arrive at a solution that didn't make this patch less robust. The main sticking point is if I switched from taskqs to an approach based on per FS thread that could potentially flood the system with new thread if, for example, a user decided to make lots of clones of a snapshot with a big unlinked set. So as a result I've left the patch mostly as is.

There are a couple of ways to deal with porting this patch to other platforms that don't have `taskq_cancel_id`. One of them is adding the missing taskq function(s) or we could drop the unlinked drain canceling on unmount part of this patch.

ahrens commented:
Thanks @alek-p .  Could you send a message to developer@open-zfs.org to give folks on other platforms a heads up about this change?  I think everyone is looking forward to this improvement, but it would be nice to give them some notice that porting it won't be totally trivial.

alek-p commented on include/sys/dsl_pool.h:
```diff
@@ -96,6 +96,7 @@ typedef struct dsl_pool {
 	struct dsl_dataset *dp_origin_snap;
 	uint64_t dp_root_dir_obj;
 	struct taskq *dp_iput_taskq;
+	struct taskq *dp_unlinked_drain_taskq;
```
> Why is this struct taskq when the addition below is taskq_t?

just following the example of iput_taskq here, but yeah that's weird...
